File: ruby-engine/lib/custom_strategy.rb
STANDARD_STRATEGIES = [
  "default",
  "userWithId",
  "gradualRolloutUserId",
  "gradualRolloutSessionId",
  "gradualRolloutRandom",
  "flexibleRollout",
  "remoteAddress",
].freeze

class CustomStrategyHandler
  def initialize
    @custom_strategies_definitions = {}
    @custom_strategy_implementations = {}
  end

  def update_strategies(json_str)
    custom_strategies = {}
    parsed_json = JSON.parse(json_str)

    features = extract_features parsed_json

    features.each do |feature|
      toggle_name = feature["name"]
      strategies = feature["strategies"]

      custom_strategies_for_toggle = strategies.select do |strategy|
        !STANDARD_STRATEGIES.include?(strategy["name"])
      end

      unless custom_strategies_for_toggle.empty?
        custom_strategies[toggle_name] = custom_strategies_for_toggle
      end
    end

    @custom_strategies_definitions = custom_strategies
  end

  def register_custom_strategies(strategies)
    strategies.each do |strategy|
      if strategy.respond_to?(:name) && strategy.name.is_a?(String) &&
         strategy.respond_to?(:enabled?)
        @custom_strategy_implementations[strategy.name] = strategy
      else
        raise "Invalid strategy object. Must have a name method that returns a String and an enabled? method."
      end
    end
  end

  def evaluate_custom_strategies(toggle_name, context)
    results = {}

    @custom_strategies_definitions[toggle_name]&.each_with_index do |strategy, index|
      key = "customStrategy#{index + 1}"
      strategy_impl = @custom_strategy_implementations[strategy["name"]]
      result = strategy_impl&.enabled?(strategy["parameters"], context) || false
      results[key] = result
    end

    results
  end

  private

  def extract_features(data)
    return data["features"] if data.is_a?(Hash) && data.key?("features")

    if data.is_a?(Hash) && data.key?("events")
      features = []

      hydration_events = data["events"].select { |e| e["type"] == "hydration" }
      features.concat(hydration_events.flat_map { |e| e["features"] || [] })

      feature_updated_events = data["events"].select { |e| e["type"] == "feature-updated" }
      features.concat(feature_updated_events.map { |e| e["feature"] })

      return features
    end

    []
  end
end

File: python-engine/yggdrasil_engine/custom_strategy.py
import json
from typing import Dict
import inspect

_STANDARD_STRATEGIES = [
    "default",
    "userWithId",
    "gradualRolloutUserId",
    "gradualRolloutSessionId",
    "gradualRolloutRandom",
    "flexibleRollout",
    "remoteAddress",
]


def get_features_json(message):
    if "features" in message:
        return message["features"]
    elif "events" in message:
        features = {}
        for event in message["events"]:
            if event["type"] == "feature-updated":
                feature = event["feature"]
                features[feature["name"]] = feature
            elif event["type"] == "feature-removed":
                features.pop(event["featureName"], None)
            elif event["type"] == "hydration":
                features = {feature["name"]: feature for feature in event["features"]}
        return list(features.values())


class CustomStrategyHandler:

    def __init__(self):
        self.strategy_definitions = {}
        self.strategy_implementations = {}

    def update_strategies(self, features_json: str):
        custom_strategies = {}
        parsed_toggles = json.loads(features_json)

        for toggle in get_features_json(parsed_toggles):
            toggle_name = toggle["name"]
            for strategy in toggle["strategies"]:
                if strategy["name"] not in _STANDARD_STRATEGIES:
                    custom_strategies[toggle_name] = strategy["name"]

            toggle_strategies = [
                strategy
                for strategy in toggle["strategies"]
                if strategy["name"] not in _STANDARD_STRATEGIES
            ]
            if toggle_strategies:
                custom_strategies[toggle_name] = toggle_strategies

        self.strategy_definitions = custom_strategies

    def register_custom_strategies(self, custom_strategies: Dict[str, any]):
        for strategy_name, strategy in custom_strategies.items():
            if hasattr(strategy, "apply"):
                apply_method = strategy.apply
                signature = inspect.signature(apply_method)
                parameters = list(signature.parameters)
                if len(parameters) != 2:
                    raise ValueError(
                        f"Custom strategy '{strategy_name}' does not have an apply method "
                        f"with exactly two parameters. Found {len(parameters)}."
                    )
                self.strategy_implementations[strategy_name] = strategy
            else:
                raise ValueError(
                    f"Custom strategy {strategy_name} does not have an apply method"
                )

    def evaluate_custom_strategies(
        self, toggle_name: str, context: dict
    ) -> Dict[str, bool]:
        results = {}
        for index, strategy in enumerate(
            self.strategy_definitions.get(toggle_name, [])
        ):
            key = f"customStrategy{index + 1}"
            strategy_impl = self.strategy_implementations.get(strategy["name"])
            result = (
                strategy_impl.apply(strategy["parameters"], context)
                if strategy_impl
                else False
            )
            results[key] = result

        return results
